k2/csrc/array_ops_inl.h

```diff
@@ -27,6 +27,14 @@
 #include <vector>
 
 namespace k2 {
+// will be used in  MaxPerSublist
+template <typename T>
+struct MaxOp {
+  __device__ T operator()(T a, T b) { return (a > b ? a : b); }
+  __device__ MaxOp() {}
```

Comment:
There are subtle differences between
```cpp
MaxOp() {}
```
and

```cpp
MaxOp() = default;
```

I would suggest
```cpp
MaxOp() = default;
MaxOp(const MaxOp &) = default;
MaxOp &operator=(const MaxOp &) = default;
```
or leave them out entirely since there is nothing to copy or initialize.

Reply:
Actually I did this intentionally, nvcc will report a warning if you use `** = default`.

```
warning: __device__ annotation is ignored on a function("MaxOp") that is explicitly defaulted on its first declaration
```
But we really need copy constructor to be device code. Of cousre do `MaxOp {}` will make the struct not POD, but it does not matter in our case as we just use MaxOp in `Reduce`.

Check here for more details https://github.com/kokkos/kokkos/issues/1473#issuecomment-488844696. Note I don't add any comments in the code as other guys could see the warning easily if they try to do ` = default`.

Reply:
Note it may be OK to not define any constructor at all (no `MaxOp() = default`, no `MaxOp() {}`), I will try this approach when really running the code.